### PR TITLE
Add custom metrics for CloudWatch

### PR DESCRIPTION
**Pull request by guikcd**
:warning: ~~Please don't merge directly, this break the creation of the environment if default not defined!~~ :warning:

## what

Enable some interesting CloudWatch metrics. I think this is considered as custom metrics (so charges can apply) because they are not created by default.

## why

Without that, it is difficult to get metrics from the ELB or instances *directly*.
Some metrics about instances such as `RootFilesystemUtil` are custom metrics but not managed by user!

## problem

~~the option need to be defined with a value, or not defined (so empty default produce an error)~~:

```
Failed Environment update activity. Reason: Configuration validation exception: Invalid option specification (Namespace: 'aws:elasticbeanstalk:healthreporting:system', OptionName: 'ConfigDocument'): Unknown configuration setting.
```

~~I don't see how to provide a way in `terraform` to define a value and let the default not defined, as it is a setting inside a single `resource` (i'm aware of interpolation conditionals but it not help here).
A solution can be to enable one metric,~~
EDIT: problem solved by creating a default JSON document.

## references

* https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/command-options-general.html
* https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/aeb-metricscollected.html

### CODE REVIEW

**Comment by osterman** [comment]
Can you supply a sample `json` doc?

**Comment by osterman** [comment]
Check out this awesome pattern using `locals`: https://github.com/cloudposse/terraform-aws-cloudfront-cdn/blob/master/examples/wordpress/main.tf

Could this help here?

e.g. `setting = "${...ternary that returns local or depending on if document is set}"`

**Comment by osterman** [comment]
Or could we set a sane default json doc that can be overridden?

**Comment by guikcd** [comment]
I didn't look at your example for now. I think as it is custom metrics, this add charges (cents) per metric added. I just would like to offer use to disable this, is possible.

**Comment by guikcd** [comment]
This is an exemple of ConfigDocument:

```
config_document = "{\"CloudWatchMetrics\": {\"Environment\": {\"InstancesOk\": 60, \"InstancesDegraded\": 60, \"ApplicationRequests4xx\": 60}}, \"Version\": 1}"
```

**Comment by osterman** [comment]
@aknysh do you have any suggestions how we can pass an optional doc for this field? 

I think it can still be done with interpolations so long as you return a proper hash/map. We've done similar things before.

**Comment by guikcd** [comment]
@osterman the problem is not interpolation but the option setting itself; if declared, can't be empty.



|Name | Description | Default | Valid Values|
| ------- | --------------- | ---------- | --------------- |
|ConfigDocument | A JSON document describing the environment and instance metrics to publish to CloudWatch. | None |

**Comment by osterman** [comment]
>  problem is not interpolation but the option setting itself; if declared, can't be empty

That's true if you're doing interpolation on `value = "${....}"`, but what I'm suggesting is to do interpolation on `setting = "${....}"`. While it seems crazy that you can do that, terraform will properly pass a `map` back to `setting` if the interpolation returns a `map`. Does that make sense?

**Comment by guikcd** [comment]
This is absolutely what i have to do, generate `setting = "${....}"` only if `json` ConfigDocument is defined.

**Comment by guikcd** [comment]
`${....}` is quite hard to generate, i have errors maybe due to lack of knowledge of Go and HCL....

**Comment by osterman** [comment]
@guikcd no problem! If you share the errors, I can probably help clear those up. Also, if you'd like, I can invite you to our slack team. Send me an email at erik@cloudposse.com if you'd like that

**Comment by guikcd** [comment]
My first problem was https://github.com/hashicorp/terraform/issues/17447:

```
root.setting[0]: not an object type for map (*ast.LiteralType)
```

By moving `setting = ` at top of list of setting, the error disappear...

File: main.tf [hunk 1]
```diff
@@ -341,6 +351,8 @@ resource "aws_elastic_beanstalk_environment" "default" {
 
   tags = "${module.label.tags}"
 
+  setting = ["${var.config_document ? local.option_config_document : local.option_empty}"]
```
Comment (osterman) [inline main.tf]: Please use explicit comparison. E.g. `var.config_document == ""`

**Comment by osterman** [comment]
Your changes LGTM. Is it working now?

**Comment by aknysh** [comment]
@guikcd  please run `terraform fmt` on TF files
https://travis-ci.org/cloudposse/terraform-aws-elastic-beanstalk-environment/builds/348508669

**Comment by guikcd** [comment]
Not yet, i think there is a type mismatch; `terraform validate` is ok, but not `terraform plan`:

```
Error: module.aws-elastic-beanstalk-environment.aws_elastic_beanstalk_environment.default: 1 error(s) occurred:

* module.aws-elastic-beanstalk-environment.aws_elastic_beanstalk_environment.default: At column 3, line 1: conditional operator cannot be used with map values in:

${var.config_document == "" ? local.option_empty : local.option_config_document}

```

**Comment by guikcd** [comment]
https://github.com/hashicorp/terraform/blob/master/vendor/github.com/hashicorp/hil/check_types.go#L445-L460 says that it is not supported by design for now.

**Comment by osterman** [comment]
Ah, that sucks! Looks like we went down a rabbit hole on that one. 

Okay, here's another blasphemous hack we can do to overcome limitations of the language. 

If `config_document` is non-empty, then use `aws:elasticbeanstalk:healthreporting:system` otherwise use `aws:elasticbeanstalk:application:environment`.

```
option_config_document {
    namespace = "${var.config_document == "" ? "aws:elasticbeanstalk:application:environment" : "aws:elasticbeanstalk:healthreporting:system"}"
    name      = "${var.config_document == "" ? "CONFIG_DOCUMENT_HACK" : "ConfigDocument"}"
    value     = "${var.config_document == "" ? "placeholder" : var.config_document}"
  }
```

Alternatively, check what the TF language does if `namespace`, `name`, and `value` are all empty? Maybe it just ignores it. That would be cleaner than defining this variable.

**Comment by osterman** [comment]
Have you tried passing an empty JSON doc that looks like this:
```
{}
```

**Comment by guikcd** [comment]
This PR is ready to be merged.

**Comment by akagr** [comment]
### Note for future visitors

Check out the [additional_settings](https://github.com/cloudposse/terraform-aws-elastic-beanstalk-environment#input_additional_settings) variable to specify custom metrics. The changes pushed in this PR aren't valid anymore.

**Pull request merged**